base/genericmemory.jl

```diff
@@ -288,3 +288,18 @@ function indcopy(sz::Dims, I::GenericMemory)
     src = eltype(I)[I[i][_findin(I[i], i < n ? (1:sz[i]) : (1:s))] for i = 1:n]
     dst, src
 end
+
+# Wrapping a memory region in an Array
+@eval function reshape(m::GenericMemory{M, T}, dims::Vararg{Int, N}) where {M, T, N}
+    len = Core.checked_dims(dims...)
+    length(m) == len || throw(DimensionMismatch("parent has $(length(m)) elements, which is incompatible with size $(dims)"))
+    ref = MemoryRef(m)
+    $(Expr(:new, :(Array{T, N}), :ref, :dims))
+end
+
+@eval function view(m::GenericMemory{M, T}, inds::AbstractUnitRange) where {M, T}
+    @boundscheck checkbounds(m, inds)
+    ref = @inbounds MemoryRef(m, first(inds)) # @inbounds is needed to allow view(Memory{T}(undef, 0), 1:0)
```

Comment:
@oscardssmith, I think this `@inbounds` is not preventing bounds checking which causes the build to fail (even though evaling these into Base in a pre-built Julia does work) . Is there a more robust way to disable bounds checking or should we avoid constructing this out of bounds `MemoryRef` entirely and branch on `isempty(inds)`?

Reply:
Since first!=1 in general, you need the branch on isempty anyways to handle that edge case

Reply:
`@inbounds` is undefined behavior if it would have thrown an error when disabled. That can lead to arbitrary memory corruption and program corruption.

Reply:
Is it possible to create an out of bounds MemoryRef at an offset, or is that always UB?

Reply:
It is considered UB to make a MemoryRef that does not reference valid memory

Reply:
This makes sense. And what makes `MemoryRef(Memory{Int64}(undef, 0))` is okay is

https://github.com/JuliaLang/julia/blob/313f933370aff4edf1eb8f972964b2e459451066/src/builtins.c#L1678-L1679

Reply:
Yes, it both does (technically) point to valid memory (size 0 memory allocate extra space for 1 element) and is special cases to be allowed for the zero-sized case to be constructed, but not computed with it

Reply:
possibly too late, but would the design have been simpler if we just disallowed 0 length memory? since it's not resizable, there isn't much need to be able to make one.

Reply:
The fact that empty memories are actually stored as length 1 memories supports that. Empty arrays can still be backed by the "one true size-one memory" of that type to avoid allocations.

Reply:
Moving this discussion to https://github.com/JuliaLang/julia/issues/53903#issue-2216508204

Reply:
There is no long `@inbounds` in the current implementation.